### PR TITLE
Add geocoding on address creation

### PR DESCRIPTION
**Pull request by dwhenry**
Address geocoding is used to populate markers at http://locator.pensionwise.gov.uk and for distance calculation within the main pension wise application.

This was missed out by accident on the previous location editing PR. As a result, the previous commit is not a valid rollback point as any edits to address corrupt the underlying data.

### CODE REVIEW

**Comment by dwhenry** [comment]
@benbarnett updated JS and fixed tests. For future reference when using VCR with Capybara and poltergiest the following config is required to avoid the weird requests beaking the test suite.
```
VCR.configure do |c|
  c.ignore_localhost = true
end
```

**Comment by benbarnett** [comment]
Ah right, good to know! 

:+1:

File: app/models/address.rb [hunk 1]
```diff
@@ -25,4 +33,22 @@ def to_a
...
+      self.point = {
+        type: 'Point',
+        coordinates: [geocode.longitude, geocode.latitude]
+      }
+    end
+  end
+
+  def geocode_postcode
+    results = Geocoder.search(postcode)
+    results.count == 1 ? results.first : nil
```
Comment (benlovell) [inline app/models/address.rb]: I'm not usually a fan of `try` but this line might be clarified by `results.try(:first)`?

File: spec/models/address_spec.rb [hunk 2]
```diff
@@ -2,15 +2,40 @@
...
+        data: { 'address_components' => [{ 'types' => ['postal_code'] }] },
+        longitude: -2.000,
+        latitude: 52.000
+      )
     end
 
-    it 'does not corrupt an existing address uid' do
-      address_1 = Address.create
-      address_2 = Address.find(address_1.id)
+    context 'when geocode look works' do
```
Comment (benlovell) [inline spec/models/address_spec.rb]: Should that be 'lookup' instead?

File: spec/models/address_spec.rb [hunk 3]
```diff
@@ -2,15 +2,40 @@
...
+
+        expect(address_1.uid).to eq(address_2.uid)
+      end
+    end
 
-      expect(address_1.uid).to eq(address_2.uid)
+    context 'when geocode lookup fails' do
+      before do
+        allow(Geocoder).to receive(:search).and_return([])
+      end
```
Comment (benlovell) [inline spec/models/address_spec.rb]: Can these be separated with a blank line? It makes them easier to scan / parse.

File: spec/services/update_location_spec.rb [hunk 4]
```diff
@@ -92,7 +92,18 @@
...
+        let(:update_params) { { address: { town: 'New town', postcode: 'AA1 2BB', address_line_1: 'Hope' } } }
+        let(:geocode) do
+          double(
+            :geocode,
+            data: { 'address_components' => [{ 'types' => ['postal_code'] }] },
+            longitude: -2.000,
+            latitude: 52.000
+          )
+        end
+        before do
```
Comment (benlovell) [inline spec/services/update_location_spec.rb]: Could we add a blank line to separate the `let`s and the `before` here too?

File: app/models/address.rb (hunk 1 above)
Comment (dwhenry) [inline app/models/address.rb]: This is designed to block the case when more than one result is returned. Which will result is the address change failing. In this case, the address will need to be manually set by us for the time being. Maybe wrapping the Geocoder with a decorator class to give it a nicer interface would make this clearer?

Comment (benlovell) [inline app/models/address.rb]: Ah, gotcha. It would be somewhat clearer if we stick this behind a declaratively named private method, as you said.

Comment (dwhenry) [inline app/models/address.rb]: @benlovell have extracted this functionality into a separate class

File: app/models/address.rb [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
-  def self.find_or_create_from_params(params)
-    find_or_create_by!(
-      address_line_1: params['address_line_1'].presence,
-      address_line_2: params['address_line_2'].presence,
-      address_line_3: params['address_line_3'].presence,
-      town: params['town'].presence,
-      county: params['county'].presence,
-      postcode: params['postcode'].presence
-    )
+  class << self
```
Comment (benlovell) [inline app/models/address.rb]: What was the rationale for switching class method declaration here? We've replaced five characters with a bunch of extra lines and noise, ultimately.

File: app/models/postcode_geocoder.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class PostcodeGeocoder
```
Comment (benlovell) [inline app/models/postcode_geocoder.rb]: Nice :+1:

File: spec/models/address_spec.rb [hunk 7]
```diff
@@ -44,4 +61,29 @@
       )
     end
   end
+
+  describe 'validations' do
+    subject { described_class.new(params).tap(&:valid?) }
```
Comment (benlovell) [inline spec/models/address_spec.rb]: Nice trick :+1:

File: app/models/address.rb (hunk 5 above)
Comment (dwhenry) [inline app/models/address.rb]: This was changed as initially I was going to have multiple class methods.
When I changed tactic I didn't revert this, I'll reverted this now.

Comment (benlovell) [inline app/models/address.rb]: Cool. We should cover this with a rubocop rule at some point in the not too distant future.

**Comment by benlovell** [comment]
Squash and merge :rocket: :heart_eyes:

**Pull request merged**